From baa6ca499f2e88afba6d43ef89ef37e5243dcd0e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Micha=C5=82=20G=C3=B3rny?= <mgorny@gentoo.org>
Date: Tue, 27 Sep 2016 09:12:51 +0200
Subject: [PATCH 01/13] [Driver] Disable OpenSUSE rules for OpenSUSE/SLES 10
 and older

Disable the OpenSUSE rules for OpenSUSE versions older than 11 as they
are incompatible with the old binutils on that distribution.

Differential Revision: https://reviews.llvm.org/D24954
---
 lib/Driver/ToolChains.cpp | 22 ++++++++++++++++++++--
 1 file changed, 20 insertions(+), 2 deletions(-)

diff --git a/lib/Driver/ToolChains.cpp b/lib/Driver/ToolChains.cpp
index 131cdae..d95e1df 100644
--- a/lib/Driver/ToolChains.cpp
+++ b/lib/Driver/ToolChains.cpp
@@ -3931,8 +3931,26 @@ static Distro DetectDistro(vfs::FileSystem &VFS) {
         .Default(UnknownDistro);
   }
 
-  if (VFS.exists("/etc/SuSE-release"))
-    return OpenSUSE;
+  File = VFS.getBufferForFile("/etc/SuSE-release");
+  if (File) {
+    StringRef Data = File.get()->getBuffer();
+    SmallVector<StringRef, 16> Lines;
+    Data.split(Lines, "\n");
+    for (const StringRef& Line : Lines) {
+      std::pair<StringRef, StringRef> SplitLine = Line.split('=');
+      int Version;
+      if (SplitLine.first.trim() == "VERSION" &&
+          !SplitLine.second.trim().getAsInteger(10, Version)) {
+        // OpenSUSE/SLES 10 and older are not supported and not compatible
+        // with our rules, so just treat them as UnknownDistro.
+        if (Version > 10)
+          return OpenSUSE;
+        else
+          return UnknownDistro;
+      }
+    }
+    return UnknownDistro;
+  }
 
   if (VFS.exists("/etc/exherbo-release"))
     return Exherbo;
-- 
2.10.1

